kythe/typescript/indexer.ts

```diff
@@ -216,18 +146,154 @@ class StandardIndexerContext implements IndexerHost {
+ * for a listing of namespace groups for various declaration types and further
+ * discussion.
+ */
+enum TSNamespace {
+  TYPE,
+  VALUE,
+  NAMESPACE,
+}
+
+/**
+ * Context represents the environment a node is declared in, and may be used for
+ * disambiguating a node's declarations if it has multiple.
+ */
+enum Context {
+  /** No disambiguation about a node's declarations. */
+  Any,
+
+  /** The node is declared as a getter. */
+  Getter,
+
+  /** The node is declared as a setter. */
+  Setter,
+}
+
+/**
+ * A SymbolVNameCache stores a mapping of symbols to the (many) VNames it may
+ * have. Each TypeScript symbol can be be of a different TypeScript namespace
+ * and be declared in a unique context, leading to a total (`TSNamespace` *
+ * `Context`) number of possible VNames for the symbol.
+ *
+ *              TSNamespace    Context
+ *              -----------    -------
+ *              TYPE           Any
+ * ts.Symbol -> VALUE       -> Getter  -> VName
+ *              NAMESPACE      Setter
+ *                             ...
+ *
+ * The `Any` context makes no guarantee of symbol declaration disambiguation.
+ * As a result, unless explicitly set for a given symbol and namespace, the
+ * VName of an `Any` context is lazily set to the VName of an arbitrary context.
+ */
+type NamespaceAndContext = string&{__brand: 'nsctx'};
+class SymbolVNameCache {
+  private readonly cache =
+      new Map<ts.Symbol, Map<NamespaceAndContext, VName>>();
+
+  /**
+   * Serializes a namespace and context as a string to lookup in the cache.
+   *
+   * Each instance of a JavaScript object is unique, so using one as a key fails
+   * because a new object would be generated every time the cache is queried.
+   */
+  private serialize(ns: TSNamespace, context: Context): NamespaceAndContext {
+    return `${ns}${context}` as NamespaceAndContext;
+  }
+
+  /** Get a symbol VName for a given namespace and context, if it exists. */
+  get(symbol: ts.Symbol, ns: TSNamespace, context: Context): VName|undefined {
+    if (this.cache.has(symbol)) {
+      const nsCtx = this.serialize(ns, context);
+      return this.cache.get(symbol)!.get(nsCtx);
+    }
+    return undefined;
+  }
+
+  /** Set a symbol VName for a given namespace and context. */
+  set(symbol: ts.Symbol, ns: TSNamespace, context: Context, vname: VName) {
+    let vnameMap = this.cache.get(symbol);
+    const nsCtx = this.serialize(ns, context);
+    if (vnameMap) {
+      // If there is already a map for the symbol, set the VName for the given
+      // namespace and context.
+      vnameMap.set(nsCtx, vname);
+    } else {
+      // If there is no map for the symbol, create a new one with an entry of
+      // the VName for the given namepsace and context.
+      this.cache.set(symbol, new Map([[nsCtx, vname]]));
+    }
+
+    // Set the symbol VName for the given namespace and `Any` context, if it has
+    // not already been set.
+    const nsAny = this.serialize(ns, Context.Any);
```

Comment:
This is only necessary if `context != Context.Any`.
It might be more performant if you included that check, although I think the logic is simpler as-is.

Reply:
I think it's okay; it's definitely more readable this way.